include/boost/hana/functional/overload_linearly.hpp

```diff
@@ -96,6 +96,11 @@ BOOST_HANA_NAMESPACE_BEGIN
             return (*this)(static_cast<F&&>(f),
                     (*this)(static_cast<G&&>(g), static_cast<H&&>(h)...));
         }
+
+        template <typename F>
+        constexpr decltype(auto) operator()(F&& f) const {
```

Comment:
If you call `overload_linearly(some-expr)` where `some-expr` is a temporary, which should often be the case because we'll use it with a lambda like `overload_linearly([](...) {...})`, you'll end up returning an rvalue reference to the temporary. Instead, and to mimic the behaviour of `overload`, you should use `hana::detail::decay` (equivalent to `std::decay`, but faster) to make sure a copy of the function is made.